File: client/util/geometry.js
export class Vector2 {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  copy() {
    return new Vector2(this.x, this.y);
  }

  map(f) {
    var out = this.copy();
    out.x = f(out.x);
    out.y = f(out.y);
    return out;
  }

  elementwise(f, other) {
    var out = this.copy();
    out.x = f(out.x, other.x);
    out.y = f(out.y, other.y);
    return out;
  }

  scale(k) {
    var out = this.copy();
    out.x *= k;
    out.y *= k;
    return out;
  }

  offset(other) {
    var out = this.copy();
    out.x += other.x;
    out.y += other.y;
    return out;
  }

  negate() {
    var out = this.copy();
    out.x = -out.x;
    out.y = -out.y;
    return out;
  }

  equals(other) {
    return this.x === other.x && this.y === other.y;
  }

  toString() {
    return "Vector2(" + this.x + ", " + this.y + ")";
  }
}

export class Vector3 {
  constructor(x, y, z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  copy() {
    return new Vector3(this.x, this.y, this.z);
  }

  map(f) {
    var out = this.copy();
    out.x = f(out.x);
    out.y = f(out.y);
    out.z = f(out.z);
    return out;
  }

  elementwise(f, other) {
    var out = this.copy();
    out.x = f(out.x, other.x);
    out.y = f(out.y, other.y);
    out.z = f(out.z, other.z);
    return out;
  }

  scale(k) {
    var out = this.copy();
    out.x *= k;
    out.y *= k;
    out.z *= k;
    return out;
  }

  offset(other) {
    var out = this.copy();
    out.x += other.x;
    out.y += other.y;
    out.z += other.z;
    return out;
  }

  negate() {
    var out = this.copy();
    out.x = -out.x;
    out.y = -out.y;
    out.z = -out.z;
    return out;
  }

  equals(other) {
    return this.x === other.x && this.y === other.y && this.z === other.z;
  }

  toString() {
    return "Vector3(" + this.x + ", " + this.y + ", " + this.z + ")";
  }
}

export class Rectangle {
  constructor(left, top, width, height) {
    this.left = left;
    this.top = top;
    this.width = width;
    this.height = height;
  }

  copy() {
    return new Rectangle(this.left, this.top, this.width, this.height);
  }

  getTopLeft() {
    return new Vector2(this.left, this.top);
  }

  getBottomRight() {
    return new Vector2(this.getRight(), this.getBottom());
  }

  getRight() {
    return this.left + this.width;
  }

  getBottom() {
    return this.top + this.height;
  }

  intersect(other) {
    var left = Math.max(this.left, other.left);
    var right = Math.min(this.getRight(), other.getRight());

    if (right <= left) {
      return null;
    }

    var top = Math.max(this.top, other.top);
    var bottom = Math.min(this.getBottom(), other.getBottom());

    if (bottom <= top) {
      return null;
    }

    return Rectangle.fromCorners(left, top, right, bottom);
  }

  contains(other) {
    return this.left <= other.left && this.getRight() >= other.getRight() &&
           this.top <= other.top && this.getBottom() >= other.getBottom();
  }

  offset(vec) {
    var out = this.copy();
    out.left += vec.x;
    out.top += vec.y;
    return out;
  }

  scale(k) {
    var out = this.copy();
    out.width *= k;
    out.height *= k;
    return out;
  }

  equals(other) {
    return this.left === other.left && this.top === other.top &&
           this.width === other.width && this.height === other.height;
  }

  toString() {
    return "Rectangle(" + this.left + ", " + this.top + ", " +
                          this.width + ", " + this.height + ")";
  }
}

Rectangle.fromCorners = (left, top, right, bottom) =>
    new Rectangle(left, top, right - left, bottom - top);

File: elpizo/util/geometry.py
class Vector2(object):
  def __init__(self, x, y):
    self.x = x
    self.y = y

  def map(self, f):
    return self.__class__(f(self.x), f(self.y))

  def elementwise(self, f, other):
    return self.__class__(f(self.x, other.x), f(self.y, other.y))

  def scale(self, k):
    return self.__class__(k * self.x, k * self.y)

  def offset(self, other):
    return self.__class__(self.x + other.x, self.y + other.y)

  def negate(self):
    return self.__class__(-self.x, -self.y)

  def __eq__(self, other):
    return self.x == other.x and self.y == other.y

  def __repr__(self):
    return "Vector2({x}, {y})".format(**self.__dict__)

  def __hash__(self):
    return hash((Vector2, self.x, self.y))


class Rectangle(object):
  def __init__(self, left, top, width, height):
    self.left = left
    self.top = top
    self.width = width
    self.height = height

  @property
  def right(self):
    return self.left + self.width

  @property
  def bottom(self):
    return self.top + self.height

  @property
  def top_left(self):
    return Vector2(self.left, self.top)

  @property
  def bottom_right(self):
    return Vector2(self.bottom, self.right)

  def intersects(self, other):
    return self.left < other.right and self.right > other.left and \
           self.top < other.bottom and self.bottom > other.top

  def contains(self, other):
    return self.left <= other.left and self.right >= other.right and \
           self.top <= other.top and self.bottom >= other.bottom

  def offset(self, vec):
    return self.__class__(self.left + vec.x, self.top + vec.y,
                     self.width, self.height)

  def scale(self, k):
    return self.__class__(self.left, self.top, k * self.width, k * self.height)

  def __eq__(self, other):
    return self.left == other.left and self.top == other.top and \
           self.width == other.width and self.height == other.height

  def __repr__(self):
    return "Rectangle({left}, {top}, {width}, {height})".format(**self.__dict__)

  def __hash__(self):
    return hash((Rectangle, self.left, self.top, self.width, self.height))
